src/ray/pubsub/subscriber.cc

```diff
@@ -18,174 +18,272 @@ namespace ray {
+
+bool Subscriber::Unsubscribe(const rpc::ChannelType channel_type,
+                             const rpc::Address &publisher_address,
+                             const std::string &message_id_binary) {
+  return Channel(channel_type)->Unsubscribe(publisher_address, message_id_binary);
+}
+
+void Subscriber::MakeLongPollingPubsubConnection(const rpc::Address &publisher_address,
+                                                 const rpc::Address &subscriber_address) {
+  const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
+  RAY_LOG(DEBUG) << "Make a long polling request to " << publisher_id;
+  auto publisher_client = publisher_client_pool_.GetOrConnect(publisher_address);
+  rpc::PubsubLongPollingRequest long_polling_request;
+  long_polling_request.mutable_subscriber_address()->CopyFrom(subscriber_address);
+
+  publisher_client->PubsubLongPolling(
+      long_polling_request, [this, publisher_address, subscriber_address](
+                                Status status, const rpc::PubsubLongPollingReply &reply) {
+        HandleLongPollingResponse(publisher_address, subscriber_address, status, reply);
+      });
+}
+
+void Subscriber::HandleLongPollingResponse(const rpc::Address &publisher_address,
+                                           const rpc::Address &subscriber_address,
+                                           const Status &status,
+                                           const rpc::PubsubLongPollingReply &reply) {
+  const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
+  RAY_LOG(DEBUG) << "Long polling request has replied from " << publisher_id;
+  auto publishers_connected_it = publishers_connected_.find(publisher_id);
+  RAY_CHECK(publishers_connected_it != publishers_connected_.end());
+  if (!status.ok()) {
+    // If status is not okay, we treat that the publisher is dead.
+    RAY_LOG(DEBUG) << "A worker is dead. subscription_failure_callback will be invoked. "
+                      "Publisher id:"
+                   << publisher_id;
+    for (const auto &channel_it : channels_) {
+      channel_it.second->HandlePublisherFailure(publisher_address);
+    }
+  } else {
+    // Otherwise, iterate on the reply and pass published messages to channels.
+    for (int i = 0; i < reply.pub_messages_size(); i++) {
+      const auto &msg = reply.pub_messages(i);
+      const auto channel_type = msg.channel_type();
+      Channel(channel_type)->HandlePublishedMessage(publisher_address, msg);
     }
   }
-  return subscription_map_.size() == 0;
+
+  if (SubscriptionExists(publisher_id)) {
+    MakeLongPollingPubsubConnection(publisher_address, subscriber_address);
+  } else {
+    publishers_connected_.erase(publisher_id);
+  }
+}
+
+inline bool Subscriber::SubscriptionExists(const PublisherID &publisher_id) {
+  for (const auto &channel_it : channels_) {
+    if (channel_it.second->SubscriptionExists(publisher_id)) {
+      return true;
+    }
+  }
+  return false;
```

Comment:
Nit: Some of these helpers could be expressed pretty easily with `std::any_of`:
```suggestion
  return std::any_of(channels_.begin(), channels_.end(), [](const auto &p) { return p.second->SubscriptionExists(publisher_id); });
```
I like to use the std algorithms helpers when I think it will make the check clearer, lessen the chance of bugs, or reduce boilerplate, but it often comes to dev personal preference. So if you prefer to keep these as explicit checks like you have above, feel free! 😁